nx/lib/nx/defn/evaluator.ex

```diff
@@ -23,64 +23,101 @@ defmodule Nx.Defn.Evaluator do
       end
 
     {fun, cache}
   end
 
-  defp eval(%Nx.Tensor{data: %Expr{op: :parameter, args: [i]}}, vars, cache) do
-    {Enum.fetch!(vars, i), cache}
+  defp eval(%Nx.Tensor{data: %Expr{op: :parameter, args: [i]}}, state, cache) do
+    {Enum.fetch!(state.vars, i), cache}
   end
 
-  defp eval(%Nx.Tensor{data: %Expr{op: :tensor, args: [t]}}, _vars, cache) do
+  defp eval(%Nx.Tensor{data: %Expr{op: :tensor, args: [t]}}, _state, cache) do
     {t, cache}
   end
 
-  defp eval(%Nx.Tensor{data: %Expr{op: :cond, args: [clauses, last]}}, vars, cache) do
-    {res, cache} = cond_clause(clauses, last, vars, cache)
-    composite_eval(res, vars, cache)
+  defp eval(%Nx.Tensor{data: %Expr{op: :cond, args: [clauses, last]}}, state, cache) do
+    {res, cache} = cond_clause(clauses, last, state, cache)
+    composite_eval(res, state, cache)
   end
 
-  defp eval(%Nx.Tensor{data: %Expr{op: :while, args: args}}, vars, cache) do
+  defp eval(%Nx.Tensor{data: %Expr{op: :while, args: args}}, state, cache) do
     [initial, _arg, condition, block] = args
-    {initial, cache} = composite_eval(initial, vars, cache)
-    {while(initial, condition, block, cache), cache}
+    {initial, cache} = composite_eval(initial, state, cache)
+    {while(initial, condition, block, state, cache), cache}
   end
 
-  defp eval(%Nx.Tensor{data: %Expr{op: :elem, args: args}}, vars, cache) do
+  defp eval(%Nx.Tensor{data: %Expr{op: :elem, args: args}}, state, cache) do
     [tuple, i, _size] = args
-    {tuple, cache} = composite_eval(tuple, vars, cache)
+    {tuple, cache} = composite_eval(tuple, state, cache)
     {elem(tuple, i), cache}
   end
 
-  defp eval(%Nx.Tensor{data: %Expr{op: :metadata, args: [expr, _meta]}}, vars, cache) do
-    eval(expr, vars, cache)
+  defp eval(%Nx.Tensor{data: %Expr{op: :attach_token, args: [token, expr]}}, state, cache) do
+    {_, cache} = eval(token, state, cache)
+    eval(expr, state, cache)
+  end
+
+  defp eval(%Nx.Tensor{data: %Expr{op: :metadata, args: [expr, _meta]}}, state, cache) do
+    eval(expr, state, cache)
   end
 
-  defp eval(%Nx.Tensor{data: %Expr{op: op, id: id}} = ans, vars, cache) do
+  defp eval(%Nx.Tensor{data: %Expr{op: op, id: id}} = ans, state, cache) do
     case cache do
       %{^id => res} -> {res, cache}
-      %{} -> eval_apply(id, op, ans, vars, cache)
+      %{} -> eval_apply(id, op, ans, state, cache)
     end
   end
 
-  defp eval(other, _vars, cache) do
+  defp eval(other, _state, cache) do
     {other, cache}
   end
 
-  defp eval_apply(id, op, ans, vars, cache) do
-    {args, cache} = Tree.apply_args(ans, cache, &eval(&1, vars, &2))
+  defp eval_apply(_id, :token, %{data: %Expr{args: [token]}}, state, cache) do
+    hooks = state.hooks
+
+    cache =
+      Enum.reduce(token.hooks, cache, fn %{callback: callback, expr: expr, name: name}, cache ->
+        cond do
+          callback = hooks[name] || callback ->
+            {expr, cache} = eval(expr, state, cache)
+            callback.(expr)
```

Comment:
```suggestion
        hook_fun = hooks[name] || callback

        cond do
          hook_fun ->
            {expr, cache} = eval(expr, state, cache)
            hook_fun.(expr)
```

Again, mostly cosmetic, but with the variable reassignment, this match seemed to have too much cognitive overload for me